extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/kernel/StageDefinition.java

```diff
@@ -212,24 +219,51 @@ public boolean doesShuffle()
 
   public boolean doesSortDuringShuffle()
   {
-    if (shuffleSpec == null) {
+    if (shuffleSpec == null || shuffleSpec.clusterBy().isEmpty()) {
       return false;
     } else {
-      return !shuffleSpec.getClusterBy().getColumns().isEmpty() || shuffleSpec.needsStatistics();
+      return shuffleSpec.clusterBy().sortable();
     }
   }
 
-  public Optional<ShuffleSpec> getShuffleSpec()
+  /**
+   * Returns the {@link ShuffleSpec} for this stage, if {@link #doesShuffle()}.
+   *
+   * @throws IllegalStateException if this stage does not shuffle
+   */
+  public ShuffleSpec getShuffleSpec()
   {
-    return Optional.ofNullable(shuffleSpec);
+    if (shuffleSpec == null) {
+      throw new IllegalStateException("Stage does not shuffle");
+    }
+
+    return shuffleSpec;
   }
```

Comment:
Is there a way to know that the stage shuffles or not before calling this method (to avoid the ISE)?

I see changes to `doesSortDuringShuffle`, but that seems to indicate more than just "does the current stage "shuffle".

Reply:
Yes: `doesShuffle`